.github/bin/run-clang-tidy.sh

```diff
@@ -76,96 +111,110 @@ fi
 # But at least there will be an error message to stderr.
-${CLANG_TIDY} --config="$(cat ${CLANG_CONFIG_FILE})" --dump-config \
-  > /dev/null 2> ${CLANG_TIDY_CONFIG_MSG}
-if [ -s ${CLANG_TIDY_CONFIG_MSG} ]; then
-  cat ${CLANG_TIDY_CONFIG_MSG}
+"${CLANG_TIDY}" --config="$(cat ${CLANG_CONFIG_FILE})" --dump-config \
+  > /dev/null 2> "${CLANG_TIDY_CONFIG_MSG}"
+if [ -s "${CLANG_TIDY_CONFIG_MSG}" ]; then
+  cat "${CLANG_TIDY_CONFIG_MSG}"
   exit 1
 fi
 
-echo ::group::Build compilation database
-
-time $(dirname $0)/make-compilation-db.sh
-
-# The files might be reported with various absolute prefixes that
-# we remove to canonicalize and report as relative paths from project
-# root.
-readonly EXEC_ROOT="$(bazel info execution_root)"
-readonly CURRENT_DIR="$(pwd)"
-readonly CANONICALIZE_SOURCE_PATH_REGEX="^\(${EXEC_ROOT}\|${CURRENT_DIR}\|\.\)/"
-
-# We create a hash of each file content to only have to look at new files.
-# (TODO: could the tidy result be different if an include content changes ?
-#  Then we have to do g++ -E (using compilation database knowing about -I etc.)
-#  or combine hashes of all mentioned #includes that are also in our list)
-# Make need to re-run file dependent on clang-tidy configuration, WORKSPACE
-# and of course file content.
-for f in $(find . -name "*.cc" -or -name "*.h" \
-             | grep -v "verilog/tools/ls/vscode" \
-           )
-do
-  (${CLANG_TIDY} --version; cat ${CLANG_CONFIG_FILE} WORKSPACE $f) \
-    | md5sum | sed -e "s|-|$f|g"
-done | sort > "${CLANG_TIDY_SEEN_CACHE}".new
-
-# Only the files with different hashes are the ones we want to process.
-join -v2 "${CLANG_TIDY_SEEN_CACHE}" "${CLANG_TIDY_SEEN_CACHE}".new \
-  | awk '{print $2}' | sort > "${FILES_TO_PROCESS}"
-
-echo "::group::$(wc -l < ${FILES_TO_PROCESS}) files to process"
-cat "${FILES_TO_PROCESS}"
-echo "::endgroup::"
+: > "${HASHES_OF_INTEREST}"      # truncate if exists; we'll re-create below
+: > "${FILE_HASHES_TO_PROCESS}"
+: > "${INCLUDE_REPLACE}"
+
+# All the files we want to run clang-tidy on: *.cc and *.h
+find . -name "*.cc" -or -name "*.h" \
+  | grep -v "verilog/tools/ls/vscode" \
+  | sed 's|^\./||' > "${FILES_OF_INTEREST}"
+
+# Changing a header should also process all files depending on it.
+# We can't do full C-preprocessing as we don't have all -I available, so
+# we simply replace references to headers with hashes of their content.
+# If content of these files change, the hashes of the includers will change.
+#   #include "foo/bar/baz.h" -> #include "a46b89dd7a7f8f108b836b50762fb6e1"
+# Ideally loop until no changes in hashes, but one level good enough.
+
+# Create a sed expression that replaces header filenames with content hash.
+for f in $(cat ${FILES_OF_INTEREST}); do
+  if [ "${f##*.}" == "h" ]; then     # only headers expected to be relevant
+    c_hash="$(cat $f | md5sum | awk '{print $1}')"
+    echo "s|$f|$c_hash|g" | sed 's|\.|\\.|g' >> "${INCLUDE_REPLACE}"
```

Comment:
Instead of the append redirection inside the loop, you could just print to stdout and then do a single redirection outside of the loop:
```
done >> "${INCLUDE_REPLACE}"
```

Reply:
Good point. Done.
This was a leftover from an earlier version of the loop that wrote into two different files.